// react-mic-waver/dist/utils/MediaStream.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMediaStream = exports["default"] = exports.MediaStreamProvider = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var MediaStreamContext = /*#__PURE__*/_react["default"].createContext({
  stream: undefined,
  start: function start() {
    return undefined;
  },
  stop: function stop() {
    return undefined;
  }
});

var useMediaStream = function useMediaStream() {
  return _react["default"].useContext(MediaStreamContext);
};

exports.useMediaStream = useMediaStream;

var MediaStreamProvider = function MediaStreamProvider(_ref) {
  var audio = _ref.audio,
      video = _ref.video,
      children = _ref.children;

  var _React$useState = _react["default"].useState(),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      stream = _React$useState2[0],
      setStream = _React$useState2[1];

  _react["default"].useEffect(function () {
    return function () {
      if (stream) {
        stream.getTracks().forEach(function (track) {
          return track.stop();
        });
      }
    };
  }, [stream]);

  var start = _react["default"].useCallback( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
    var mediaStream;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return navigator.mediaDevices.getUserMedia({
              audio: audio,
              video: video
            });

          case 2:
            mediaStream = _context.sent;
            setStream(mediaStream);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  })), [audio, video]);

  var stop = _react["default"].useCallback(function () {
    if (stream) {
      stream.getTracks().forEach(function (track) {
        return track && track.stop();
      });
      setStream(undefined);
    }
  }, [stream]);

  return /*#__PURE__*/_react["default"].createElement(MediaStreamContext.Provider, {
    value: {
      stream: stream,
      start: start,
      stop: stop
    }
  }, children);
};

exports.MediaStreamProvider = MediaStreamProvider;
var _default = MediaStreamProvider;
exports["default"] = _default;
